### PR TITLE
webrender: run event_loop in main thread.

### LINKED ISSUE #366: thread '<unnamed>' panicked at 'On macOS, `EventLoop` must be created on the main thread!'
```
thread '<unnamed>' panicked at 'On macOS, `EventLoop` must be created on the main thread!', /Users/declan/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.24.0/src/platform_impl/macos/event_loop.rs:60:17
stack backtrace:
   0: std::panicking::begin_panic
   1: winit::event_loop::EventLoop<T>::with_user_event
   2: webrender::output::Output::create_webrender_window::{{closure}}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

### PR DESCRIPTION
**Pull request by harryfei**
Fixes #366.

### CODE REVIEW

**Comment by declanqian** [comment]
Quick notes here.
I did a quick test for this branch on Linux. There are two issues

- Not working on Wayland. Have to launch a graphic UI with `WINIT_UNIX_BACKEND=x11 ./src/emacs -Q`
- Multi frame crashes with
```
thread '<unnamed>' panicked at 'swap_buffers: eglSwapBuffers failed (eglGetError returned 0x300d)', ~/.cargo/registry/src/github.com-1ecc6299db9ec823/glutin-0.26.0/src/api/egl/mod.rs:606:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Fatal error 6: Aborted
```

**Comment by harryfei** [comment]
1. I developed it mainly on wayland. So there is some bugs which is not existed in my envrionment.
2. I'm not try to fix the multi frame supporting issue in this branch. Need additional works to make it workable.

**Comment by harryfei** [comment]
Can you post more information to help me find root cause of issue 1?

**Comment by declanqian** [comment]
There aren't much information.
After I typed `WINIT_UNIX_BACKEND=wayland ./src/emacs -Q`,
my terminal got stuck and `C-c` is not working either.
Also, no Emacs frame is visible.

Below are some output from GDB

```
[declan@thinkpadx1extreme emacs]$ WINIT_UNIX_BACKEND=wayland gdb --args ./src/emacs -Q
GNU gdb (GDB) 10.2
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./src/emacs...
(gdb) r
Starting program: /home/declan/Source/emacs/src/emacs -Q
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
[New Thread 0x7ffff3ccb640 (LWP 70655)]
[New Thread 0x7ffff1203640 (LWP 70656)]
[New Thread 0x7ffff0a02640 (LWP 70657)]
[New Thread 0x7fffdffff640 (LWP 70658)]
[New Thread 0x7fffdf7fe640 (LWP 70659)]
[New Thread 0x7fffdeffd640 (LWP 70660)]



^C
Thread 1 "emacs" received signal SIGINT, Interrupt.
0x00007ffff727c92e in epoll_wait () from /usr/lib/libc.so.6
(gdb) Quit
(gdb)
```

**Comment by harryfei** [comment]
can you try to add logs in `wr_select` in event_loop.rs and `Output` in output.rs and debug them?

**Comment by declanqian** [comment]
Ok, I'll try to do that.

**Comment by declanqian** [comment]
I built `emacs --with-webrender --without-javascript` on macOS.
I am expecting **`emacs -nw`** to work as expected, however I am getting this error.
```
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 22, kind: InvalidInput, message: "Invalid argument" }', crates/webrender/src/event_loop.rs:205:49
```

**Comment by harryfei** [comment]
I don't have a mac device, so I can't debug any macOS issue.

**Comment by declanqian** [comment]
> Not working on Wayland. Have to launch a graphic UI with WINIT_UNIX_BACKEND=x11 ./src/emacs -Q

This issue can solved by commenting out this line
https://github.com/harryfei/emacs-ng/blob/main-thread-loop/rust_src/crates/webrender/src/output.rs#L89
as suggested by @harryfei 

Not sure why it is there in the first place.
It's certainly not working with my titling WM(sway).

Also, tracking the macOS specific issue in [another thread here](https://github.com/emacs-ng/emacs-ng/pull/242).

**Comment by harryfei** [comment]
I add a 100ms timeout in `wait_for_window_resize`.  I used it to get a stable window size.

**Comment by declanqian** [comment]
Does that 100ms got added to `emacs-init-time`?

**Comment by harryfei** [comment]
> Does that 100ms got added to `emacs-init-time`?

It's here.
https://github.com/harryfei/emacs-ng/blob/7c202f3ee0c4097a4a2abeaac78b678b29c2aabe/rust_src/crates/webrender/src/event_loop.rs#L53

**Comment by declanqian** [comment]
I meant does it slow down Emacs initialization.
PGTK Emacs `(emacs-init-time)` in my setup is 0.132407 seconds,
while WR Emacs is 0.455714 seconds.

Anyway, this is another issue.

**Comment by harryfei** [comment]
of course, there will be a 0.1s delay at worst case.

File: rust_src/crates/webrender/src/event_loop.rs [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+    }
+}
+
+pub static EVENT_LOOP: Lazy<Mutex<WrEventLoop>> = Lazy::new(|| {
+    Mutex::new(WrEventLoop {
+        el: glutin::event_loop::EventLoop::with_user_event(),
+    })
+});
+
+pub static TOKIO_RUNTIME: Lazy<Mutex<Runtime>> =
```
Comment (DavidDeSimone) [inline rust_src/crates/webrender/src/event_loop.rs:104]: One thing I thing looking at this - would there be benefit to having a shared Tokio runtime between JS and the event loop if both js and WebRender are enabled?

**Review by DavidDeSimone — APPROVED** [review]
This is a lot of code, but overall it looks good to me and so think is a good addition.

File: rust_src/crates/webrender/src/event_loop.rs (hunk 1 above)
Comment (harryfei) [reply rust_src/crates/webrender/src/event_loop.rs:104]: Using a shared tokio runtime will make emacs-ng spawn less threads.

**Pull request merged**